File: Yogscart Alpha/Assets/Scripts/KartMaker Scripts/KartMaker.js
#pragma strict

//enums
enum KartType{Display,Local,Online,Spectator};

private var gd : CurrentGameData;
private var im : InputManager;

var DebugMode : boolean;

function Start()
{

gd = transform.GetComponent(CurrentGameData);
im = transform.GetComponent(InputManager);

	if(DebugMode)
		SpawnKart(KartType.Local,Vector3.zero,Quaternion.identity,Random.Range(0,gd.Karts.Length),Random.Range(0,gd.Wheels.Length),Random.Range(0,gd.Characters.Length),Random.Range(0,gd.Hats.Length));

}

function SpawnKart(kartType : KartType, position : Vector3, rotation : Quaternion, kart : int, wheel : int, character : int, hat : int) : Transform
{

	//Spawn Kart & Wheels
	var kartBody : Transform = Instantiate(gd.Karts[kart].model,Vector3(0,0,0),Quaternion.identity);
	var kartSkel : KartSkeleton = kartBody.GetComponent(KartSkeleton);

	var frontlWheel : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.FrontLPosition,Quaternion.Euler(0,0,0));
	frontlWheel.parent = kartBody.FindChild("Kart Body");
	frontlWheel.name = "FrontL Wheel";

	var frontrWheel : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.FrontRPosition,Quaternion.Euler(0,180,0));
	frontrWheel.parent = kartBody.FindChild("Kart Body");
	frontrWheel.name = "FrontR Wheel";

	var backlWheel : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.BackLPosition,Quaternion.Euler(0,180,0));
	backlWheel.parent = kartBody.FindChild("Kart Body");
	backlWheel.name = "BackL Wheel";

	var backrWheel : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.BackRPosition,Quaternion.Euler(0,0,0));
	backrWheel.parent = kartBody.FindChild("Kart Body");
	backrWheel.name = "BackR Wheel";

	//Spawn Character & Hat
	var characterMesh : Transform = Instantiate(gd.Characters[character].model,Vector3(0,0,0),Quaternion.identity);
	characterMesh.name = "Character";

	var charSkel : CharacterSkeleton = characterMesh.GetComponent(CharacterSkeleton);

	characterMesh.position = kartSkel.SeatPosition - charSkel.SeatPosition;
	characterMesh.parent = kartBody.FindChild("Kart Body");

	if(hat != 0){

	var hatMesh : Transform = Instantiate(gd.Hats[hat].Model,charSkel.HatHolder.position,Quaternion.identity);
	hatMesh.parent = charSkel.HatHolder;
	hatMesh.localRotation = Quaternion.Euler(0,0,0);

	}

	if(kartType != KartType.Display){

	var kb : GameObject = kartBody.gameObject;

	kb.AddComponent(Rigidbody);
	kb.GetComponent.<Rigidbody>().mass = 2000;
	kb.GetComponent.<Rigidbody>().collisionDetectionMode = CollisionDetectionMode.ContinuousDynamic;
	kb.GetComponent.<Rigidbody>().interpolation = RigidbodyInterpolation.Interpolate;
	kb.GetComponent.<Rigidbody>().angularDrag = 0;

	kb.AddComponent(AudioSource);
	kb.GetComponent.<AudioSource>().clip = kartSkel.engineSound;
	kb.GetComponent.<AudioSource>().spatialBlend = 1;
	kb.GetComponent.<AudioSource>().minDistance = 0;
	kb.GetComponent.<AudioSource>().maxDistance = 35;
	kb.GetComponent.<AudioSource>().rolloffMode = AudioRolloffMode.Linear;
	
	var kartBodyBody = kartBody.FindChild("Kart Body").gameObject;
	kartBodyBody.AddComponent(AudioSource);
	kartBodyBody.GetComponent.<AudioSource>().spatialBlend = 1;
	kartBodyBody.GetComponent.<AudioSource>().minDistance = 0;
	kartBodyBody.GetComponent.<AudioSource>().maxDistance = 25;
	kartBodyBody.GetComponent.<AudioSource>().rolloffMode = AudioRolloffMode.Linear;
	
	kb.GetComponent.<AudioSource>().playOnAwake = false;
	kartBodyBody.GetComponent(AudioSource).playOnAwake = false;

	kb.AddComponent(DeathCatch);
	kb.GetComponent(DeathCatch).DeathParticles = kartBody.FindChild("Kart Body").FindChild("Particles").FindChild("Death Particles").GetComponent.<ParticleSystem>();

	var frontlWheelCollider : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.FrontLPosition,Quaternion.Euler(0,0,0));
	frontlWheelCollider.name = "FrontL Wheel";
	frontlWheelCollider.parent = kartBody.FindChild("Colliders");
	SetUpWheelCollider(frontlWheelCollider);

	var frontrWheelCollider : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.FrontRPosition,Quaternion.Euler(0,180,0));
	frontrWheelCollider.parent = kartBody.FindChild("Colliders");
	frontrWheelCollider.name = "FrontR Wheel";
	SetUpWheelCollider(frontrWheelCollider);

	var backlWheelCollider : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.BackLPosition,Quaternion.Euler(0,180,0));
	backlWheelCollider.parent = kartBody.FindChild("Colliders");
	backlWheelCollider.name = "BackL Wheel";
	SetUpWheelCollider(backlWheelCollider);

	var backrWheelCollider : Transform = Instantiate(gd.Wheels[wheel].model,kartSkel.BackRPosition,Quaternion.Euler(0,0,0));
	backrWheelCollider.parent = kartBody.FindChild("Colliders");
	backrWheelCollider.name = "BackR Wheel";
	SetUpWheelCollider(backrWheelCollider);

	kb.AddComponent(kartAnimation);
	kb.GetComponent(kartAnimation).ani = characterMesh.GetComponent(Animator);

	kb.AddComponent(kartScript);

	var ks = kb.GetComponent(kartScript);

	ks.engineSound = kartSkel.engineSound;

	ks.wheelColliders = new WheelCollider[4];
	ks.wheelColliders[0] = frontlWheelCollider.GetComponent(WheelCollider);
	ks.wheelColliders[1] = frontrWheelCollider.GetComponent(WheelCollider);
	ks.wheelColliders[2] = backlWheelCollider.GetComponent(WheelCollider);
	ks.wheelColliders[3] = backrWheelCollider.GetComponent(WheelCollider);

	ks.wheelMeshes = new Transform[4];
	ks.wheelMeshes[0] = frontlWheel;
	ks.wheelMeshes[1] = frontrWheel;
	ks.wheelMeshes[2] = backlWheel;
	ks.wheelMeshes[3] = backrWheel;

	var kp : Transform = kartBody.FindChild("Kart Body").FindChild("Particles");

	ks.flameParticles = new ParticleSystem[2];
	ks.flameParticles[0] = kp.FindChild("L_Flame").GetComponent.<ParticleSystem>();
	ks.flameParticles[1] = kp.FindChild("R_Flame").GetComponent.<ParticleSystem>();

	ks.DriftParticles = new Transform[2];
	ks.DriftParticles[0] = kp.FindChild("L_Sparks");
	ks.DriftParticles[1] = kp.FindChild("R_Sparks");

	ks.TrickParticles = kp.FindChild("Trick").GetComponent.<ParticleSystem>();

	kb.AddComponent(Position_Finding);
	kb.AddComponent(kartUpdate);
	
	kb.GetComponent(kartScript).hitSounds = gd.Characters[character].hitSounds;
	kb.GetComponent(kartScript).tauntSounds = gd.Characters[character].tauntSounds;


	if(kartType != KartType.Display)
	{
		if(kartType != KartType.Spectator)
		{
			//Add Script
			kb.AddComponent(kartInput);
			kb.AddComponent(kartInfo);
			//Adjust Scripts
			kb.GetComponent(kartInput).InputNum = 0;
		}
		//Apply to Spectators
		kb.AddComponent(kartItem);
		kb.GetComponent(kartItem).itemDistance = kartSkel.ItemDrop;
		
		}
	}

	//Clear Up
	Destroy(kartSkel);
	Destroy(charSkel);

	kartBody.position = position;
	kartBody.rotation = rotation;
	
	kartBody.gameObject.layer = 8;//Set the Kart's Layer to "Kart" for Kart Collisions 
	
	return kartBody.transform;

}

function SetUpWheelCollider(collider : Transform)
{

collider.gameObject.AddComponent(WheelCollider);

var wheelCollider = collider.GetComponent(WheelCollider);

//Setup Collider Settings
wheelCollider.mass = 20;

if(collider.GetComponent(WheelSkeleton) != null)
{
	wheelCollider.radius = collider.GetComponent(WheelSkeleton).wheelRadius;
	Destroy(collider.GetComponent(WheelSkeleton));
}
else
	wheelCollider.radius = 0.15;

wheelCollider.wheelDampingRate = 0.05;
wheelCollider.suspensionDistance = 0.25;
wheelCollider.forceAppPointDistance = 1.25f;

wheelCollider.suspensionSpring.spring = 25000;
wheelCollider.suspensionSpring.damper = 25000;
wheelCollider.suspensionSpring.targetPosition = 1;

wheelCollider.forwardFriction.extremumSlip = 0.8;
wheelCollider.forwardFriction.extremumValue = 3;
wheelCollider.forwardFriction.asymptoteSlip = 1.5;
wheelCollider.forwardFriction.asymptoteValue = 2.25;
wheelCollider.forwardFriction.stiffness = 1;

wheelCollider.sidewaysFriction.extremumSlip = 0.8;
wheelCollider.sidewaysFriction.extremumValue = 3;
wheelCollider.sidewaysFriction.asymptoteSlip = 1.5;
wheelCollider.sidewaysFriction.asymptoteValue = 2.25;
wheelCollider.sidewaysFriction.stiffness = 3;

Destroy(collider.GetComponent(MeshFilter));
Destroy(collider.GetComponent(MeshRenderer));


}

File: Yogscart C#/Assets/Scripts/Kart Maker/KartMaker.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using UnityEngine.Networking;

public enum KartType { Display, Local, Online, Spectator };

public class KartMaker : MonoBehaviour
{

    static CurrentGameData gd;

    void Start()
    {
        gd = transform.GetComponent<CurrentGameData>();
    }

    public Transform SpawnKart(KartType type, Vector3 pos, Quaternion rot, int c, int h, int k, int w)
    {
        //Spawn Kart and Wheels
        Transform kartBody = Instantiate(gd.karts[k].model, Vector3.zero, Quaternion.identity);
        Transform actualKartBody = kartBody.Find("Kart Body"); ;

        Kart_Skeleton kartSkel = kartBody.GetComponent<Kart_Skeleton>();

        Transform frontlWheel, frontrWheel, backlWheel, backrWheel;

        frontlWheel = Instantiate(gd.wheels[w].model, kartSkel.FrontLPosition, Quaternion.Euler(0, 0, 0));
        frontlWheel.parent = actualKartBody;
        frontlWheel.name = "FrontL Wheel";

        frontrWheel = Instantiate(gd.wheels[w].model, kartSkel.FrontRPosition, Quaternion.Euler(0, 180, 0));
        frontrWheel.parent = actualKartBody;
        frontrWheel.name = "FrontR Wheel";

        backlWheel = Instantiate(gd.wheels[w].model, kartSkel.BackLPosition, Quaternion.Euler(0, 0, 0));
        backlWheel.parent = actualKartBody;
        backlWheel.name = "BackL Wheel";

        backrWheel = Instantiate(gd.wheels[w].model, kartSkel.BackRPosition, Quaternion.Euler(0, 180, 0));
        backrWheel.parent = actualKartBody;
        backrWheel.name = "BackR Wheel";

        //Spawn Character and Hat
        Transform characterMesh = Instantiate(gd.characters[c].model, Vector3.zero, Quaternion.identity);
        characterMesh.name = "Character";

        Character_Skeleton charSkel = characterMesh.GetComponent<Character_Skeleton>();
        characterMesh.position = kartSkel.SeatPosition - charSkel.SeatPosition;
        characterMesh.parent = actualKartBody;

        if (h != 0)//Don't spawn a hat if hat value equals zero
        {
            Transform hatMesh = Instantiate(gd.hats[h].model, charSkel.HatHolder.position, Quaternion.identity);
            hatMesh.parent = charSkel.HatHolder;
            hatMesh.localRotation = Quaternion.Euler(0, 0, 0);
        }

        //Setup IK Animation
        DrivingIK ikComponent = characterMesh.gameObject.AddComponent<DrivingIK>();
        ikComponent.leftHandTarget = kartSkel.leftHandTarget;
        ikComponent.rightHandTarget = kartSkel.rightHandTarget;
        ikComponent.leftFootTarget = kartSkel.leftFootTarget;
        ikComponent.rightFootTarget = kartSkel.rightFootTarget;
        ikComponent.steeringWheel = actualKartBody.Find("Steering Wheel");

        if (type == KartType.Online)
        {
            kartBody.gameObject.AddComponent<NetworkIdentity>();
        }

        if (type != KartType.Display)
        {
            GameObject kb = kartBody.gameObject;

            //Setup Rigidbody
            Rigidbody rb = kb.AddComponent<Rigidbody>();
            rb.mass = 1000;
            rb.collisionDetectionMode = CollisionDetectionMode.ContinuousDynamic;
            rb.interpolation = RigidbodyInterpolation.Interpolate;
            rb.angularDrag = 0.05f;

            //Set up Kart sounds
            AudioSource aus = kb.AddComponent<AudioSource>();
            aus.clip = kartSkel.engineSound;
            aus.spatialBlend = 1;
            aus.minDistance = 0f;
            aus.maxDistance = 35f;
            aus.rolloffMode = AudioRolloffMode.Linear;
            aus.playOnAwake = false;

            AudioSourceInfo ausI = kb.AddComponent<AudioSourceInfo>();
            ausI.idealVolume = 0.05f;
            ausI.audioType = AudioSourceInfo.AudioType.SFX;

            //Set up Kart Body Sounds
            AudioSource kbas = actualKartBody.gameObject.AddComponent<AudioSource>();
            kbas.spatialBlend = 1;
            kbas.minDistance = 0f;
            kbas.maxDistance = 100f;
            kbas.rolloffMode = AudioRolloffMode.Linear;
            kbas.playOnAwake = false;
            kbas.dopplerLevel = 0f;
            kbas.spread = 0f;

            ausI = kbas.gameObject.AddComponent<AudioSourceInfo>();
            ausI.idealVolume = 1f;
            ausI.audioType = AudioSourceInfo.AudioType.SFX;

            //Setup Wheel Colliders
            Transform frontLWheelCollider = new GameObject().transform;
            frontLWheelCollider.position = kartSkel.FrontLPosition;
            frontLWheelCollider.name = "Front L Wheel";
            frontLWheelCollider.parent = kartBody.Find("Colliders");
            SetUpWheelCollider(frontLWheelCollider);

            Transform frontRWheelCollider = new GameObject().transform;
            frontRWheelCollider.name = "Front R Wheel";
            frontRWheelCollider.position = kartSkel.FrontRPosition;
            frontRWheelCollider.parent = kartBody.Find("Colliders");
            SetUpWheelCollider(frontRWheelCollider);

            Transform backLWheelCollider = new GameObject().transform;
            backLWheelCollider.position = kartSkel.BackLPosition;
            backLWheelCollider.name = "Back L Wheel";
            backLWheelCollider.parent = kartBody.Find("Colliders");
            SetUpWheelCollider(backLWheelCollider);

            Transform backRWheelCollider = new GameObject().transform;
            backRWheelCollider.position = kartSkel.BackRPosition;
            backRWheelCollider.name = "Back R Wheel";
            backRWheelCollider.parent = kartBody.Find("Colliders");
            SetUpWheelCollider(backRWheelCollider);

            //Add Kart Script
            KartMovement km = kb.AddComponent<KartMovement>();

            km.characterID = c;
            km.engineSound = kartSkel.engineSound;

            //Spawn Particles
            Transform particlesPack = Instantiate(Resources.Load<Transform>("Prefabs/Kart Maker/Bodies/Particles"), actualKartBody.position, actualKartBody.rotation, actualKartBody);

            km.particleSystems = new Dictionary<string, ParticleSystem>();

            string[] particleNames = new string[] { "Death Particles", "L_Flame", "R_Flame", "L_Sparks", "R_Sparks", "Trick", "L_DriftClouds", "R_DriftClouds", "L_StartClouds", "R_StartClouds" };
            int count = 0;
            foreach (string particle in particleNames)
            {
                //Add Particle System to dictioanry of kart's particles
                Transform currentparticle = particlesPack.Find(particle);
                km.particleSystems.Add(particle, currentparticle.GetComponent<ParticleSystem>());

                switch (count)
                {
                    case 0: currentparticle.position = kartSkel.deathParticlesPos; break;
                    case 1: currentparticle.position = kartSkel.rFlame; break;
                    case 2: currentparticle.position = kartSkel.lFlame; break;
                    case 3: currentparticle.position = kartSkel.rSparks; break;
                    case 4: currentparticle.position = kartSkel.lSparks; break;
                    case 5: currentparticle.position = kartSkel.trick; break;
                    case 6: currentparticle.position = kartSkel.lDriftClouds; break;
                    case 7: currentparticle.position = kartSkel.rDriftClouds; break;
                    case 8: currentparticle.position = kartSkel.lStartClouds; break;
                    case 9: currentparticle.position = kartSkel.rStartClouds; break;
                }

                count++;
            }

            //Add Death Catch
            kb.AddComponent<DeathCatch>();
            kb.GetComponent<DeathCatch>().deathParticles = km.particleSystems["Death Particles"];

            kb.AddComponent<CowTipping>();

            //Add Other Kart Scripts
            kb.AddComponent<PositionFinding>();
            //Sort out Character Noises

            //Add Animator Script
            kb.AddComponent<kartAnimator>();
            kb.GetComponent<kartAnimator>().ani = characterMesh.GetComponent<Animator>();

            if (type != KartType.Spectator)
            {
                kb.AddComponent<KartInput>();
                //kb.AddComponent<kartInfo>();
            }

            kb.AddComponent<KartItem>();
            kb.GetComponent<KartItem>().itemDistance = kartSkel.ItemDrop;

            kb.AddComponent<KartCollider>();
        }

        Destroy(kartSkel);
        Destroy(charSkel);

        kartBody.position = pos;
        kartBody.rotation = rot;

        kartBody.gameObject.layer = 8;//Set the Kart's Layer to "Kart" for Kart Collisions 

        return kartBody.transform;

    }

    void SetUpWheelCollider(Transform collider)
    {
        FauxCollider wheelCollider = collider.gameObject.AddComponent<FauxCollider>();

        wheelCollider.springCoefficent = 40f;
        wheelCollider.dampeningCoefficent = 7f;

        SphereCollider sphere = collider.gameObject.AddComponent<SphereCollider>();
        sphere.radius = 0.06f;
    }
}